File: FloodFill_LC733.java
/* for both dfs and dfs
Time: O(m*n)
Space: O(m*n)
*/

class Solution {
    public int[][] floodFill(int[][] image, int sr, int sc, int newColor) {
        if(image == null || image.length == 0 || 
           image[0].length == 0 || image[sr][sc] == newColor)
            return image;
        
        Queue<Integer> q = new LinkedList<>();
        int oldColor = image[sr][sc];
        int n = image.length;
        int m = image[0].length;
        image[sr][sc] = newColor;
        q.add(sr); q.add(sc);
        
        
        int[][] dirs = {{0,1},{1,0},{-1,0},{0,-1}};
        
        while(!q.isEmpty()){
            int r = q.poll();
            int c = q.poll();
            
            for(int[] dir: dirs){
                int nr = dir[0] + r;
                int nc = dir[1] + c;
                
                if(nr >= 0 && nr < n && nc >= 0 && nc < m && image[nr][nc] == oldColor){
                    image[nr][nc] = newColor;
                    q.add(nr);
                    q.add(nc);
                }
            }
        }
        return image;
    }
}

class Solution2 {
    int m; int n; int oldColor;
    int[][] dirs = {{0,1},{1,0},{-1,0},{0,-1}};
    public int[][] floodFill(int[][] image, int sr, int sc, int newColor) {
        if(image == null || image.length == 0 || 
           image[0].length == 0 || image[sr][sc] == newColor)
            return image;
        
        
        oldColor = image[sr][sc];
        n = image.length;
        m = image[0].length;
        
        dfs(image, sr, sc, newColor);
        return image;
    }
    
    private void dfs(int[][] image, int sr, int sc, int newColor){
        //base
        if(sr < 0 || sr >= n || sc < 0 || sc >= m || image[sr][sc] != oldColor)
            return;
        //logic
        image[sr][sc] = newColor;
        for(int[] dir: dirs){
            int r = sr + dir[0];
            int c = sc + dir[1];
            dfs(image, r, c, newColor);
        }
    }
}

File: floodFill-LC 733.py

# BFS
# Time Complexity = O(m * n)
# Space Complexity = O(m * n)

# 2 queues solution
class Solution:
    def floodFill(self, image: list[list[int]], sr: int, sc: int, color: int) -> list[list[int]]:
        # If the cell already has the new color, then return the image as it is
        if image == None or len(image) == 0 or image[sr][sc] == color:
            return image
        
        m = len(image)
        n = len(image[0])
        
        # Directions array - up, down, left, right
        dirs = [[-1, 0], [1, 0], [0, -1], [0, 1]]
        
        # 2 queues declaration
        rows = []
        cols = []
        
        # Capture the old color
        oldColor = image[sr][sc]
        
        rows.append(sr)
        cols.append(sc)
        
        # Change the first cell to new color
        image[sr][sc] = color
        
        
        while rows:
            currR = rows.pop(0)
            currC = cols.pop(0)
            for d in dirs:
                nr = currR + d[0]
                nc = currC + d[1]
                if (nr >= 0 and nr < m and nc >= 0 and nc < n and image[nr][nc] == oldColor):
                    rows.append(nr)
                    cols.append(nc)
                    image[nr][nc] = color
                    
                    
                    
        return image
     




# DFS Solution
# Time Complexity = O(m * n)
# Space Complexity = O(m * n)

class Solution:
    def floodFill(self, image: list[list[int]], sr: int, sc: int, color: int) -> list[list[int]]:
        # If the cell already has the new color, then return the image as it is
        if image == None or len(image) == 0 or image[sr][sc] == color:
            return image
        
        self.m = len(image)
        self.n = len(image[0])
        
        # Directions array - up, down, left, right
        self.dirs = [[-1, 0], [1, 0], [0, -1], [0, 1]]
        
        # Capture the old color
        self.oldColor = image[sr][sc]
        
        self.dfs(image, sr, sc, self.oldColor, color)
        
        return image
    
    
    def dfs(self, image, row, col, oldColor, color):
        # Base case
        # If cell is not of the old color, return
        if row < 0 or row == self.m or col < 0 or col == self.n or image[row][col] != self.oldColor:
            return
        
        # Logic
        # Change the first cell to new color
        image[row][col] = color
        
        for d in self.dirs:
            nr = row + d[0]
            nc = col + d[1]
            self.dfs(image, nr, nc, oldColor, color)
                
                
        